Add pynest version of hpc benchmark

jakobj opened the pull request:
This PR adds a pynest version of `hpc_benchmark.sli` to enable systematic benchmarking also via the Python interface.
I tried to stay as close as possible to the original implementation.
I propose @mschmidt87 and @heplesser as reviewers.

heplesser commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+
+import numpy as np
+import os
+# from scipy.special import lambertw
```
Please remove if not needed.

heplesser commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+'''
+ Parameter section
+
+ define all relevant parameters: changes should be made here
```
Define

heplesser commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+    'presimtime': 50.,      # simulation time until reaching equilibrium
+    'dt': 0.1,              # simulation step
+    'record_spikes': True,  # switch to record spikes of excitatory neurons to file
+    'path_name': './',      # path where all files will have to be written
```
I think the default path name (CWD) should be empty, since the expression for "this directory" may differ from `.` on non-Unix systems. The slash should definitely not be included.

heplesser commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+
+
+def convert_synapse_weight(tau_m, tau_syn, C_m):
+    '''Computes conversion factor for synapse weight from mV to pA. This
```
Start docstrings with a brief single-line description, then details on new line. Here, break after "pA."

heplesser commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+# evoked post-synaptic potential is 1.700759 ms.
+# For alpha-shaped postynaptic currents, the rise time of the post-synaptic
+# potential follows from the synaptic rise time as
+
```
Add # here to make clear that this is one long comment; same on line 125.

heplesser commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+    while i < len(nodes):
+        if nest.GetStatus([nodes[i]], 'local')[0]:
+            yield nodes[i]
+            i += nvp
```
This is only correct if we have nodes with proxies, isn't it? That definitely needs to be documented in the docstring.

jakobj replied:
I think it should also work for nodes without proxies. The minimal gid separation is nvp, isn't it? And in case that node is not local, we just start incrementing by one until we find a local node. At least that was the logic I had in mind. Updated the docstring.

jakobj replied:
I see now. You are right. This will only work for nodes with proxies. The minimal step size for nodes without proxies could be one.

mschmidt87 commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+   This benchmark was originally developed for very large-scale simulations on
+   supercomputers with more than 1 million neurons in the network and
+   11.250 incoming synapses per neuron. For such large networks, input to a
+   single neuron will be little correlated and network activity will remain
```
Correlated across neurons, right?
Perhaps change to "synaptic inputs will have low correlations across neurons"

jakobj replied:
Yes, I think this means across neurons. This is the text from `hpc_benchmark.sli`. Should I also change it there?

mschmidt87 replied:
I would say, yes, though it is a minor point.

mschmidt87 commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+   supercomputers with more than 1 million neurons in the network and
+   11.250 incoming synapses per neuron. For such large networks, input to a
+   single neuron will be little correlated and network activity will remain
+   stable for long periods of time.
```
stable over long periods of time.

jakobj replied:
(see above)

mschmidt87 commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+
+def convert_synapse_weight(tau_m, tau_syn, C_m):
+    '''Computes conversion factor for synapse weight from mV to pA. This
+    function is specific to the used neuron model Leaky
```
to the used neuron model which is a leaky ...

mschmidt87 commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+'''
+
+
+def build_network(logger):
```
Add a brief description of the function incl. description of function argument here, something like

'''Build the network including setting of simulation and neuron parameters, creation of neurons and connections.

Parameters
---------------
logger : Instance of Logger class
'''

mschmidt87 commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+
+    nest.message(M_INFO, 'build_network', 'Creating inhibitory population.')
+    I_neurons = nest.Create('iaf_psc_alpha', NI)  # subnet gets own gid
+
```
What is the purpose of this comment? You are not creating a subnet here, are you?

mschmidt87 commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+    logger.log(str(memory_thisjob()) + ' # virt_mem_after_edges')
+
+
+def run_simulation():
```
Same as above: Please add a brief description of the function.

mschmidt87 commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+def compute_rate():
+    '''compute approximation of average firing rate based on number of
+    local nodes, number of local spikes and total time
+    '''
```
There should be one common format for docstrings of functions across the script.
Either like this:
'''...
'''

or 
'''
...
'''

mschmidt87 commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+            # convert rank to string, prepend 0 if necessary to make
+            # numbers equally wide for all ranks
+            rank = '{:0' + str(len(str(self.max_rank_log))) + '}'
+            fn = file_name + '_' + rank.format(nest.Rank()) + '.dat'
```
It would be better to use proper string formatting here. For instance:

`fn = "{}_{}.dat".format(file_name, rank.format(nest.Rank))`

The same for the `log`-Function below.

jakobj commented:
Thank you for the quick review. I've now implemented almost all suggested changes. I've also changed the logger to a context manager, to make sure file are properly closed also on aborted simulations.

jakobj commented:
Fixed the formatting and adapted doc according to @mschmidt87's suggestions.

heplesser commented on pynest/examples/hpc_benchmark.py:
```diff
@@ -0,0 +1,487 @@
+#     b = 1.0 / tau_syn - 1.0 / tau_m
+#
+# Inverting this equation numerically leads to tau_syn =
+# 0.32582722403722841 ms, as specified in model_params below.
```
@mschmidt87 I am a bit confused here, and since you inserted the corresponding comments/code in the SLI version of the benchmark a while ago, I hope you can clarify what is happening here. My first concern is that in the function `PSP_rise_time()` above a line seems to be missing after the two lines defining `a` and `b`---maybe the line defining `t_max` on line 111 above? As it is written now, there is nothing really to invert.

The other question is why we want to achieve the rise time of 1.700759 ms?

heplesser replied:
@mschmidt87 I am missing a line in the sli and the py version of hpc_benchmark. The `PSP_rise_time` function, as far as I can see, only sets local variables `a` and `b`, but does not compute any rise time and does not return any value. I *assume* that the rise time the function is supposed to compute is the `t_max` expression in function `ConvertSynapseWeight`. Probably someone here planned to split that function up, forgot to copy a line, and then never followed quite through. If you agree with this interpretation, we could change both versions consistently.
The duplicate definition of `/tau_syn` looks like a code duplication introduced by a merge.

mschmidt87 replied:
Yes, you are absolutely right, I agree to your interpretation, sorry for being a bit slow-witted here. Line 111 should be moved inside `PSP_rise_time`, the code then be commented in, and lines 108 and 111 can then be replaced by a call of `PSP_rise_time` (`b` is still needed in the rest of the function). Also, `t_max` can be renamed to `t_rise`, or at least we should settle on either "time of maximum" or "rise time".

jakobj commented:
With the PRs of @hakonsbm and @heplesser merged into my branch, this should be ready to go to master, please have another look

heplesser commented:
@jakobj There seem to be a few PEP8 issues left, could you take a look?

jakobj commented:
Done
